File: clients/android/app/src/main/kotlin/com/lextures/android/core/notebook/NotebookStore.kt
package com.lextures.android.core.notebook

import android.content.Context
import android.util.Base64
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.time.Instant
import java.util.UUID

/** Mirrors web `student-notebook-storage` (device-local notebooks, format v2). */
@Serializable
data class NotebookPage(
    val id: String,
    val title: String = "Untitled",
    val parentId: String? = null,
    val sortOrder: Int = 0,
    val kind: String = "page",
    val contentMd: String = "",
) {
    companion object {
        fun new(title: String = "Untitled", sortOrder: Int = 0): NotebookPage =
            NotebookPage(id = UUID.randomUUID().toString(), title = title, sortOrder = sortOrder)
    }
}

@Serializable
data class CourseNotebook(
    val formatVersion: Int = 2,
    val updatedAt: String = "",
    val courseTitle: String? = null,
    val pages: List<NotebookPage> = emptyList(),
    val activePageId: String? = null,
) {
    val previewText: String
        get() = pages.joinToString("\n\n") { it.contentMd }.trim()

    companion object {
        fun empty(): CourseNotebook {
            val page = NotebookPage.new()
            return CourseNotebook(
                updatedAt = Instant.now().toString(),
                pages = listOf(page),
                activePageId = page.id,
            )
        }
    }
}

/** Device-local notebook persistence keyed per signed-in user (JWT `sub`), parity with web localStorage. */
class NotebookStore(context: Context, accessToken: String?) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val ownerKey = jwtSubject(accessToken) ?: "anonymous"
    private val json = Json { ignoreUnknownKeys = true }

    private val storageKey: String
        get() = "lextures.studentNotebooks.v1:$ownerKey"

    private fun readAll(): Map<String, CourseNotebook> {
        val raw = prefs.getString(storageKey, null) ?: return emptyMap()
        return runCatching { json.decodeFromString<Map<String, CourseNotebook>>(raw) }.getOrDefault(emptyMap())
    }

    private fun writeAll(notebooks: Map<String, CourseNotebook>) {
        prefs.edit().putString(storageKey, json.encodeToString(notebooks)).apply()
    }

    fun load(courseCode: String): CourseNotebook = readAll()[courseCode] ?: CourseNotebook.empty()

    fun save(courseCode: String, notebook: CourseNotebook) {
        val all = readAll().toMutableMap()
        all[courseCode] = notebook.copy(updatedAt = Instant.now().toString())
        writeAll(all)
    }

    fun exists(courseCode: String): Boolean = readAll().containsKey(courseCode)

    /** Every stored course code, including the global key. */
    fun allCourseCodes(): List<String> = readAll().keys.toList()

    /** Write a server copy verbatim — keeps the server `updatedAt` so last-write-wins stays stable. */
    fun saveFromServer(courseCode: String, notebook: CourseNotebook) {
        val all = readAll().toMutableMap()
        all[courseCode] = notebook
        writeAll(all)
    }

    /** Course-scoped notebooks with content (excludes the global key). */
    fun listCourseNotebooks(): Map<String, CourseNotebook> =
        readAll().filter { (key, notebook) -> key != GLOBAL_KEY && notebook.previewText.isNotEmpty() }

    companion object {
        /** Learner-wide notebook key — must not collide with real course codes (same value as web). */
        const val GLOBAL_KEY = "__lextures_global__"
        const val GLOBAL_TITLE = "Global notebook"
        private const val PREFS_NAME = "lextures_student_notebooks"

        fun jwtSubject(token: String?): String? {
            if (token.isNullOrBlank()) return null
            val parts = token.split(".")
            if (parts.size < 2) return null
            return runCatching {
                val payload = Base64.decode(parts[1], Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
                val obj = Json.parseToJsonElement(String(payload)) as? JsonObject ?: return null
                obj["sub"]?.jsonPrimitive?.content?.takeIf { it.isNotEmpty() }
            }.getOrNull()
        }
    }
}

File: clients/ios/Lextures/Core/Notebook/NotebookStore.swift
import Foundation

/// Mirrors web `student-notebook-storage` (device-local notebooks, format v2).
struct NotebookPage: Codable, Identifiable, Hashable {
    var id: String
    var title: String
    var parentId: String?
    var sortOrder: Int
    var kind: String // "page" | "group"
    var contentMd: String

    static func new(title: String = "Untitled", sortOrder: Int = 0) -> NotebookPage {
        NotebookPage(
            id: UUID().uuidString.lowercased(),
            title: title,
            parentId: nil,
            sortOrder: sortOrder,
            kind: "page",
            contentMd: ""
        )
    }
}

struct CourseNotebook: Codable {
    var formatVersion: Int
    var updatedAt: String
    var courseTitle: String?
    var pages: [NotebookPage]
    var activePageId: String?

    static func empty() -> CourseNotebook {
        let page = NotebookPage.new()
        return CourseNotebook(
            formatVersion: 2,
            updatedAt: ISO8601DateFormatter().string(from: Date()),
            courseTitle: nil,
            pages: [page],
            activePageId: page.id
        )
    }

    var previewText: String {
        pages.map(\.contentMd).joined(separator: "\n\n").trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

/// Device-local notebook persistence keyed per signed-in user (JWT `sub`), parity with web localStorage.
@MainActor
final class NotebookStore {
    /// Learner-wide notebook key — must not collide with real course codes (same value as web).
    static let globalKey = "__lextures_global__"
    static let globalTitle = "Global notebook"

    private let defaults = UserDefaults.standard
    private let ownerKey: String

    init(accessToken: String?) {
        ownerKey = NotebookStore.jwtSubject(from: accessToken) ?? "anonymous"
    }

    private var storageKey: String {
        "lextures.studentNotebooks.v1:\(ownerKey)"
    }

    private func readAll() -> [String: CourseNotebook] {
        guard let data = defaults.data(forKey: storageKey) else { return [:] }
        return (try? JSONDecoder().decode([String: CourseNotebook].self, from: data)) ?? [:]
    }

    private func writeAll(_ notebooks: [String: CourseNotebook]) {
        guard let data = try? JSONEncoder().encode(notebooks) else { return }
        defaults.set(data, forKey: storageKey)
    }

    func load(courseCode: String) -> CourseNotebook {
        readAll()[courseCode] ?? .empty()
    }

    func save(courseCode: String, notebook: CourseNotebook) {
        var all = readAll()
        var next = notebook
        next.updatedAt = ISO8601DateFormatter().string(from: Date())
        all[courseCode] = next
        writeAll(all)
    }

    func exists(courseCode: String) -> Bool {
        readAll()[courseCode] != nil
    }

    /// Every stored course code, including the global key.
    func allCourseCodes() -> [String] {
        Array(readAll().keys)
    }

    /// Write a server copy verbatim — keeps the server `updatedAt` so last-write-wins stays stable.
    func saveFromServer(courseCode: String, notebook: CourseNotebook) {
        var all = readAll()
        all[courseCode] = notebook
        writeAll(all)
    }

    /// Course-scoped notebooks with content (excludes the global key).
    func listCourseNotebooks() -> [String: CourseNotebook] {
        readAll().filter { key, notebook in
            key != NotebookStore.globalKey && !notebook.previewText.isEmpty
        }
    }

    static func jwtSubject(from token: String?) -> String? {
        guard let token else { return nil }
        let parts = token.split(separator: ".")
        guard parts.count >= 2 else { return nil }
        var payload = String(parts[1])
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while payload.count % 4 != 0 { payload.append("=") }
        guard
            let data = Data(base64Encoded: payload),
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let sub = json["sub"] as? String, !sub.isEmpty
        else { return nil }
        return sub
    }
}
